"레스토랑 상세 페이지 구현하기" 제출

anpaul0615 opened the pull request.

anpaul0615 committed:
Restaurant 관련 화면을 RestaurantPage 로 분리

anpaul0615 committed:
App 컴포넌트에 라우팅 적용

anpaul0615 committed:
HomePage 컴포넌트 테스트 작성 및 구현

anpaul0615 committed:
AboutPage 컴포넌트 테스트 작성 및 구현

anpaul0615 committed:
NotFoundPage 컴포넌트 테스트 작성 및 구현

anpaul0615 committed:
App 컴포넌트에 header 추가

anpaul0615 committed:
RestaurantPage 라우팅 경로 수정

anpaul0615 committed:
HomePage 컴포넌트 라우팅 목록 수정

anpaul0615 committed:
AboutPage, NotFoundPage 테스트코드에서 MemoryRouter 제거

anpaul0615 committed:
service/api.fetchRestaurant 테스트 작성 및 구현

anpaul0615 committed:
actions.getRestaurant, actions.setRestaurant 테스트 작성 및 구현

anpaul0615 committed:
reducer.selectRestaurant, reducer.setRestaurant 테스트 작성 및 구현

anpaul0615 committed:
RestaurantContainer 테스트 작성 및 구현

anpaul0615 committed:
RestaurantViewPage 테스트 작성 및 구현

anpaul0615 committed:
RestaurantPage → RestaurantListPage 명칭 변경

anpaul0615 committed:
RestaurantContainer 테스트에서 menuItems[].id  누락부분 수정

anpaul0615 committed:
레스토랑 상세보기 관련 action 수정

- selectRestaurant, getRestaurant 제거
- getRestaurantById 추가

anpaul0615 committed:
reducer 에서 selectRestaurant 제거

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,48 @@
+  }
+
+  beforeEach(() => {
+    useDispatch.mockImplementation(() => dispatch);
```
여기 위에 `dispatch.mockClear()`를 넣어야 혹시 모를 문제에 대해서 방지할 수 있습니다.

anpaul0615 replied:
네.. 제가 목업 관련 초기화를 놓쳤군요..
얼른 고쳐놓겠습니다..!

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,48 @@
+    }));
+  });
+
+  it('with path /', () => {
```
무엇을 테스트하는지 이름에서 드러나지 않는 것 같아요. 어떤걸 기대하고 있는건가요?

anpaul0615 replied:
describe 를 혼동해서 적었군요..!
지적감사합니다. 얼른 고쳐놓을게요.

hannut91 commented on src/actions.js:
```diff
@@ -67,3 +68,30 @@ export function loadRestaurants() {
   };
 }
+
+export function selectRestaurant(restaurantId) {
+  return {
+    type: 'selectRestaurant',
+    payload: { restaurantId },
+  };
+}
+
```
이 액션은 언제 사용되나요?

anpaul0615 replied:
아직 푸시를 안해서 저 액션을 사용하는 코드를 못보셨을것같습니다.

처음 의도는 특정 레스토랑이 클릭됐을때, 그 선택된 레스토랑을 store 에서 관리하도록 하기위해서 저 액션을 만들었습니다. 물론 이 의도대로라면 관련 리듀서와 컨테이너들도 비슷한 방향대로 구현됐습니다.

그런데 이 방법은 완전히 잘못된 방법이었습니다. 
상세보기 페이지 렌더링할때 store 에 있는 값을 사용하게되면, 브라우저를 새로고침하면 렌더링이 안되니깐요..!

그래서 레스토랑 상세보기 관련 action 을 완전히 뜯어 고치게 됩니다..ㅠㅠ

처음엔 force push 를 생각했었는데.. 
그래도 이왕 삽질한거 그대로 남겨보자는 생각으로 열심히 수정커밋 만들었습니다.

이제는 상세보기 페이지가 브라우저 URL 을 사용해서 렌더링됩니다..^^

anpaul0615 committed:
RestaurantViewPage 리팩토링

- 레스토랑 아이디를 가져오는 방식 변경
- 레스토랑 상세보기 관련 action 변경

anpaul0615 committed:
RestaurantContainer 에 레스토랑 상세보기 페이지 링크 적용

anpaul0615 committed:
RestaurantContainer 메뉴목록 표출 예외처리

anpaul0615 committed:
RestaurantViewPage 라우팅 설정

anpaul0615 committed:
App 컴포넌트 테스트 명세 수정

anpaul0615 committed:
App 컴포넌트 테스트에 dispatch.mockClear() 추가

anpaul0615 committed:
e2e 테스트 동작오류 수정

- HomePage 에서 "Home" 소제목을 찾는 테스트케이스
  - HomePage 에 소제목 추가
- HomePage 에서 "Restaurants" 링크목록을 찾는 테스트케이스
  - HomePage 내 링크목록 명칭을 "레스토랑" 에서 "Restaurants" 로 변경
- 곧바로 /restaurants/1 로 접근하는 테스트케이스
  - index.html 의 main.js 를 절대경로로 변경

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,66 @@
+      regions: [],
+      categories: [],
+      restaurants: [],
+      restaurant: {
+        name: '레스토랑 이름',
+        address: '레스토랑 주소',
+        menuItems: [
+          { id: 1, name: '메뉴 이름' },
+        ],
+      },
```
fixture를 만들었으니 fixture를 활용해도 좋을 것 같아요

anpaul0615 replied:
그러게요. 왜 하드코딩했을까요.. ㅎㅎㅎ 수정해서 반영하겠습니다^^

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,66 @@
+  describe('with path /about', () => {
+    it('return About Page', () => {
+      const { queryByText } = renderComponent({ path: '/about' });
+      expect(queryByText('헤더')).not.toBeNull();
```
지금은 헤더만 확인하고 있어서 정확히 이 테스트가 무엇을 하는지 확인하기 어려운 것 같아요. about 페이지라는 것을 알 수 있도록 테스트를 작성하는 것은 어떨까요?

anpaul0615 replied:
생각해보니깐 `pages/*.test` 에서 각 페이지 렌더링 여부 및 렌더링 결과를 테스트하고 있네요.
굳이 `App.test` 에서 렌더링 결과를 다시 확인할필요는 없어보여서 `renderComponent({ path })` 로 각 페이지의 렌더링 여부만 확인하도록 해보겠습니다.

hannut91 commented on src/RestaurantContainer.test.jsx:
```diff
@@ -0,0 +1,27 @@
+
+test('RestaurantContainer', () => {
+  useSelector.mockImplementation((selector) => selector({
+    restaurant: {
+      name: '레스토랑 이름',
+      address: '레스토랑 주소',
+      menuItems: [
+        { id: 1, name: '메뉴 이름' },
+      ],
+    },
```
여기도 `fixture`를 사용하면 좋을 것 같네요

anpaul0615 replied:
반영하겠습니다^^

hannut91 commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,24 @@
+
+import { get } from './utils';
+
+export default function RestaurantContainer() {
```
이 컴포넌트는 `useSelector`를 사용해서 `redux`를 알고 잇고 그리고 화면에 어떻게 그릴지도 알고 있는 두 가지 일을 하고 있네요. 우리가 2강에서 배웠던 컴포넌트로 관심사의 분리를 해보는건 어떨까요?

anpaul0615 replied:
의견주신대로 Container 과 Presentational 로 관심사를 분리해보겠습니다.

hannut91 commented on src/actions.test.js:
```diff
@@ -83,4 +85,36 @@ describe('actions', () => {
+      });
+    });
+
+    context('without restaurantId', () => {
```
`with null`이 더 적절할 것 같아요

anpaul0615 replied:
그러네요. `describe` 가 `getRestaurantById` 이니깐, `context` 는 전달받는 파라미터 상태를 표현하게되니, 문맥상 `with null` 로도 충분히 읽히네요. 
좋은 의견 감사합니다^^

hannut91 commented on src/reducer.test.js:
```diff
@@ -108,4 +111,24 @@ describe('reducer', () => {
+        restaurant: {},
+      };
+
+      const restaurant = {
+        name: '레스토랑 이름',
+        address: '레스토랑 주소',
+        menuItems: [
+          { name: '메뉴 이름' },
+        ],
+      };
```
여기도 fixture를 사용하는건 어떨까요?

anpaul0615 replied:
반영하겠습니다^^

hannut91 commented on index.html:
```diff
@@ -6,6 +6,6 @@
   <body>
     <div id="app"></div>
-    <script src="main.js"></script>
+    <script src="/main.js"></script>
```
절대경로로 바꾸겨서 문제를 해결하셨네요. index.html의 `<head>`에 `<base href="/" />`를 입력하여 해결할 수도 있습니다  

See also
  - https://developer.mozilla.org/en-US/docs/Web/HTML/Element/base

anpaul0615 replied:
오..! 이런방법이..!!

anpaul0615 replied:
사실 절대경로로 수정한게 엄청 찜찜했었는데.. 덕분에 굉장히 명퀘한 해결방법을 배운것같습니다.
감사합니다^^

anpaul0615 committed:
RestaurantListPage, RestaurantViewPage 에 소제목 추가

anpaul0615 committed:
App 컴포넌트 테스트를 페이지 렌더링 여부만 확인하도록 변경

anpaul0615 committed:
action.getRestaurantById() 테스트 명세 수정

anpaul0615 committed:
*.test 내부 하드코딩되어있는 값을 fixture 로 교체

anpaul0615 committed:
소스코드 관심사별 디렉토리 구분

anpaul0615 committed:
LinkList 컴포넌트 추가

anpaul0615 committed:
RestaurantsContainer 리팩토링

- Presentational 관심사를 LinkList 컴포넌트로 위임

anpaul0615 committed:
List 컴포넌트 추가

anpaul0615 committed:
RestaurantViewPanel 컴포넌트 추가

anpaul0615 committed:
RestaurantsContainer 리팩토링

- Presentational 관심사를 RestaurantViewPanel 컴포넌트로 위임

anpaul0615 committed:
ButtonList 컴포넌트 추가

anpaul0615 committed:
CategoriesContainer 리팩토링

- Presentational 관심사를 ButtonList 컴포넌트로 위임

anpaul0615 committed:
RegionsContainer 리팩토링

- Presentational 관심사를 ButtonList 컴포넌트로 위임

anpaul0615 committed:
index.html 수정

- base path 를 / 로 지정
- main.js 를 상대경로로 재변경

anpaul0615 committed:
e2e 테스트 동작오류 수정

- store 경로 수정

hannut91 commented on src/components/ButtonList.test.jsx:
```diff
@@ -0,0 +1,56 @@
+
+import ButtonList from './ButtonList';
+
+import CATEGORIES from '../../fixtures/categories';
```
ButtonList에서는 CATEGORIES랑 완전히 상관없어야해요.

anpaul0615 replied:
오 정말 동의합니다..! 
props 는 items 이라고 해놓고 테스트에서는 CATEGORIES 를 사용했네요.
바로 교정해볼게요!

hannut91 commented on src/components/LinkList.test.jsx:
```diff
@@ -0,0 +1,36 @@
+import LinkList from './LinkList';
+
+describe('<LinkList />', () => {
+  const renderComponent = (restaurants) => render((
+    <MemoryRouter>
+      <LinkList items={restaurants} />
+    </MemoryRouter>
+  ));
```
여기도 restaurants라는 이름이 등장하면 안될 것 같아요

anpaul0615 replied:
바로 수정해볼게요!

hannut91 commented on src/components/LinkList.test.jsx:
```diff
@@ -0,0 +1,36 @@
+    </MemoryRouter>
+  ));
+
+  const itemsWithLink = [
+    { id: 1, name: '아이템 이름 1', link: '아이템 링크 1' },
+    { id: 2, name: '아이템 이름 2', link: '아이템 링크 2' },
+    { id: 3, name: '아이템 이름 3', link: '아이템 링크 3' },
+  ];
```
`links`라는 이름으로도 충분할 것 같아요

anpaul0615 replied:
제안 감사해요ㅎㅎ 수정해볼게요!

hannut91 commented on src/components/List.test.jsx:
```diff
@@ -0,0 +1,32 @@
+import List from './List';
+
+describe('<List />', () => {
+  const renderComponent = (restaurants) => render((
+    <List items={restaurants} />
+  ));
```
여기도 restaurants라는 이름이 등장하면 안될 것 같아요

anpaul0615 replied:
바로 수정해볼게요!

hannut91 commented on src/store/reducer.test.js:
```diff
@@ -63,49 +67,55 @@ describe('reducer', () => {
+    it('changes restaurant', () => {
+      const initialState = {
+        restaurant: {},
+      };
+
+      const restaurant = RESTAURANT;
+
+      const state = reducer(initialState, setRestaurant(restaurant));
 
-      expect(state.selectedCategory).toEqual({
-        id: 1,
-        name: '한식',
-      });
+      expect(state.restaurant).toEqual(restaurant);
     });
```
저는 아직 `const restaurant = RESTAURANT`의 의미를 잘 모르겠어요. 저라면 이렇게 할 것 같아요

```js
describe('setRestaurant', () => {
  const initialState = {
    restaurant: {},
  };

  it('changes restaurant', () => {
    const state = reducer(initialState, setRestaurant(RESTAURANT));

    expect(state.restaurant).toEqual(RESTAURANT);
  });
});
```

anpaul0615 replied:
불필요한 asign 을 줄이자는 말이죠?
예전 리뷰에서도 말씀드렸었는데 코드 모양을 동일하게 맞추려고 하다보니 `const restaurant = RESTAURANT;` 를 했던것같아요. 

그런데 다시생각해보니 테스트 읽는데 쓸데없는 생각("저는 아직 const restaurant = RESTAURANT의 의미를 잘 모르겠어요. " 와 같은) 이 들것같아요. 

이부분은 고쳐볼게요.

hannut91 commented on src/store/reducer.test.js:
```diff
@@ -63,49 +67,55 @@ describe('reducer', () => {
       const initialState = {
-        categories: [
-          { id: 1, name: '한식' },
-        ],
+        categories: CATEGORIES,
         selectedCategory: null,
       };
 
-      const state = reducer(initialState, selectCategory(1));
+      const selectedCategory = CATEGORIES[0];
+
+      const state = reducer(initialState, selectCategory(selectedCategory.id));
+
+      expect(state.selectedCategory).toEqual(selectedCategory);
+    });
+  });
```
초기 상태에서 왜 categories를 가지고 있는지 잘 모르겠어요 저라면 이렇게 할 것 같아요

```js
describe('selectCategory', () => {
  const categoryId = 1;

  const initialState = {
    selectedCategory: null,
  };

  it('changes selected category', () => {
    const state = reducer(initialState, selectCategory(categoryId));

    expect(state.selectedCategory).toEqual(categoryId);
  });
});
```

anpaul0615 replied:
왜냐하면 이번과제 초기 코드베이스에서 `selectCategory` 카테고리아이디만 관리하는게 아니라 선택된 카테고리 객체 자체를 관리하기 때문이죠. 아래 코드는 제가 구현한게 아닌 초기 저장소를 fork/clone 했을때부터 있었던 리듀서 코드입니다.

```
  selectCategory(state, { payload: { categoryId } }) {
    const { categories } = state;
    return {
      ...state,
      selectedCategory: categories.find(equal('id', categoryId)),
    };
  },
```

사실 저는 윤석님 의도처럼 카테고리를 선택했을때 아이디만 관리하도록 구현했었는데, 아샬님은 객체를 관리하도록 구현했더라구요.

anpaul0615 replied:
그런이유로 댓글로 올려주신 테스트는 현재 리듀서 코드에서 동작하지않아요.
지금 코드에서의 리듀서는 `categories.find(equal('id', categoryId))` 를 하기때문에 `categories` 가 꼭 필요하거든요.

hannut91 replied:
아하 그렇군요. categories가 필요하네요. 제가 구현한 코드랑 헷갈렸습니다.

anpaul0615 replied:
넵넵 ㅎㅎ 사실 저도 윤석님이랑 같은 생각이었어요..!

hannut91 replied:
역시 코드리뷰로는 설계에 대한 리뷰에 한계가 있습니다. 이런 것은 짝 프로그래밍을 하면 좋아요

hannut91 commented on src/store/reducer.test.js:
```diff
@@ -63,49 +67,55 @@ describe('reducer', () => {
-        ],
+        regions: REGIONS,
         selectedRegion: null,
       };
 
-      const state = reducer(initialState, selectRegion(1));
+      const selectedRegion = REGIONS[0];
+
+      const state = reducer(initialState, selectRegion(selectedRegion.id));
 
-      expect(state.selectedRegion).toEqual({
-        id: 1,
-        name: '서울',
-      });
+      expect(state.selectedRegion).toEqual(selectedRegion);
     });
   });
```
저라면 이렇게 할 것 같아요

```js
describe('selectRegion', () => {
  const regionId = 1;

  const initialState = {
    selectedRegion: null,
  };

  it('changes selected region', () => {
    const state = reducer(initialState, selectRegion(regionId));

    expect(state.selectedRegion).toEqual(regionId);
  });
});
```

anpaul0615 replied:
위의 `describe('selectCategory', () => {})` 와 동일한 이유입니다^^

anpaul0615 committed:
reducer.test 에서 불필요한 asign 제거

anpaul0615 committed:
List, LinkList, ButtonList 에서 도메인 의존성 제거

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,65 @@
+  });
+
+  describe('with any page', () => {
+    it('have Header', () => {
```
헤더가 렌더링 되는 것을 테스트하고 있는 것 같아요. 테스트 대상이 무엇을 하는지 명시해주면 좋을 것 같아요. `renders header`는 어떨까요?

anpaul0615 replied:
말씀들어보니 테스트 의도가 제대로 전달되지 않는것같아요.
의견반영해서 수정해보겠습니다^^

anpaul0615 replied:
`renders header` 보다 `has header` 가 좀 더 적절해보여요ㅎㅎ

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,34 +1,65 @@
-      { id: 1, name: '마법사주방' },
-    ],
-  }));
+  describe('with path /', () => {
+    it('return Home Page', () => {
+      renderComponent({ path: '/' });
+    });
+  });
```
Home Page가 렌더 되는 것을 테스트하고 있는 것 같아요. `return`이 아닌 `renders Home Page`라고 하고 Home Page를 테스트 중이라는 것을 더 드러내는 것은 어떨까요?

anpaul0615 replied:
수정해보겠습니다..!

anpaul0615 replied:
테스트 의도를 명확하게 드러내는데 실패한것같네요 ㅎㅎ

hannut91 commented on src/components/ButtonList.jsx:
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+
+export default function ButtonList({
+  listId = 'button-list', items, selectedItem, handleClickItem,
```
버튼을 클릭했을 때 어떤 동작을 할지 props로 넘겨주는 것 같아요. `handle`이라는 이름은 실제 이벤트를 처리하는 곳에서 사용하는 이름이고 여기서는 다른 이름을 사용하는게 좋을 것 같아요. 실제 함수를 사용하는 곳을 기준으로 이름을 짓는게 좋을 것 같아요. onClick은 어떨까요?

anpaul0615 replied:
음.. 생각해보니깐 실제 함수를 사용하는 곳에서는 그게 어떤함수인지 몰라야할것같아요.
단순하게 클릭이벤트가 일어났을때 어떤 함수가 호출되어야하는지만 연결되면 될것같아요.
의견주신대로 onClick~ 으로 시작하는 함수명이 좋을것같아요.

anpaul0615 replied:
`onClickButton` 으로 함수명을 수정하려고했는데, 생각해보니깐 이 컴포넌트(ButtonList) 자체가 Button 의 목록을 표출하는거라서 `onClick` 만으로도 `Button 에 대한 Click` 이라고 문맥상 읽힐것같아요.

hannut91 commented on src/components/ButtonList.test.jsx:
```diff
@@ -0,0 +1,56 @@
+    });
+
+    context('when click category-button', () => {
+      it('fired click event', () => {
```
`fires`가 맞을 것 같아요

anpaul0615 replied:
it 에 들어가는 명세는 3인칭 현재형 동사로 시작해야되는거죠?
수정해보겠습니다..!

hannut91 commented on src/components/ButtonList.test.jsx:
```diff
@@ -0,0 +1,56 @@
+
+import ITEMS from '../../fixtures/items';
+
+describe('<ButtonList />', () => {
```
`ButtonList`는 카테고리랑 아무런 상관이 없어야 할 것 같아요.

anpaul0615 replied:
`fixture` 는 수정했는데 테스트명세나 변수명을 수정안했었네요..!
얼른 고치겠습니다.

hannut91 commented on src/components/RestaurantViewPanel.test.jsx:
```diff
@@ -0,0 +1,31 @@
+    });
+  });
+
+  context('without restaurant', () => {
+    it('display empty view-panel', () => {
+      const { container } = renderComponent({});
+      expect(container).not.toHaveTextContent(RESTAURANT.name);
+      expect(container).not.toHaveTextContent(RESTAURANT.address);
+      expect(container).not.toHaveTextContent(RESTAURANT.menuItems[0].name);
+    });
+  });
```
컴포넌트로 아무런 값을 넘겨주지 않았는데 RESTAURANT이 렌더링 되지 않는다고 테스트 하는 것은 조금 어색한 것 같아요. 비어있다는 것을 어떻게 표현하면 좋을까요?

anpaul0615 replied:
컴포넌트가 표출할 정보가 없는 상태를 처리하도록 수정했습니다..!

hannut91 commented on src/components/RestaurantViewPanel.test.jsx:
```diff
@@ -0,0 +1,31 @@
+
+import RESTAURANT from '../../fixtures/restaurant';
+
+describe('<RestaurantViewPanel />', () => {
```
`ViewPanel`이 어떤 의미를 가지고 있나요? `ViewPanel`이 없으면 이 컴포넌트를 이해하는데 어려울까요?

anpaul0615 replied:
containers/RestaurantContainer 와 매칭되는 components/Restaurant 로 명칭을 바꿔봤습니다.

anpaul0615 committed:
App 컴포넌트 테스트 명세 표현 개선

anpaul0615 committed:
ButtonList 컴포넌트의 handleClickItem 명칭을 onClick 으로 변경

anpaul0615 committed:
RestaurantViewPanel 컴포넌트 명칭을 Restaurant 으로 변경

anpaul0615 committed:
Restaurant 컴포넌트에 데이터 누락 예외상황 처리 추가

anpaul0615 committed:
e2e 테스트 동작오류 수정

- 레스토랑이 선택되지 않은 상태를 {} 에서 null 로 변경

hannut91 reviewed (approved):
고생하셨습니다 ㅎㅎ 다양한 시도들이 좋네요. 다양한 시도를 하면서 어떤게 더 좋은지 비교해가면서 학습하시면 좋을 것 같아요. 실수를 할 수도 있는데 실수는 리뷰어가 같이 검토해드리니 마음껏 실수하셔도 괜찮을 것 같아요. 그리고 아샬님의 과제풀이 영상을 통해 꼭 복습해보세요!

이번 과제를 통해 무엇을 배우셨나요? 더 배우고 싶은게 있으신가요?

anpaul0615 commented:
이번주차에서는 컴포넌트와 컨테이너의 관심사분리에 많은 고민을 했었습니다. 컴포넌트 명칭부터 props 까지 세세하게 도메인 의존성을 분리해내는 과정을 경험했습니다.

- 이벤트 처리 함수명칭 handle** 과 on** 를 컴포넌트 역할에 따라 구분하기
- 컴포넌트 명칭 정제하기

Presentation 컴포넌트는 화면처리만 관리하고 Container 컴포넌트는 상태만 관리하도록 하기위해서, 각자의 관심사가 아닌것들을 하나하나씩 제거해나갔는데, 그러다보니 점점 테스트코드가 매우 독립적으로 정제되어진다는걸 알게되었습니다.

- 불필요한 asign 줄이기
- 테스트에서 도메인 의존성 제거하기
- 테스트 명세에 테스트 의도를 더 정확하게 표현하기

앞으로 다음과제를 진행하면서 계속 코드가 변하게될텐데, 그럴때마다 이번과제에서 배웠던 관심사 분리를 떠올리면서 코드를 더 읽기쉽고 군더더기없도록 만들고싶습니다. 파이팅!

hannut91 commented:
관심사의 분리에 대해서 더 깊게 이해하게 되셨군요 ㅎㅎㅎ 관심사의 분리는 프로그래밍이라는 분야 전체에서 굉장히 아주 주요한 주제이기 때문에 여기서 배운 원리를 다른 곳에서도 적용해보시면 좋을 것 같아요.😀
